### PR TITLE
Double maximum length of HTTP requests

### LINKED ISSUE #5761: scripts/copydocs.py fails when the received content is too large
When the copydocs script is used to copy a work that contains data that exceeds a character count of 4094, an error is thrown and the work is not copied.

### Steps to Reproduce

1. Run the copydocs script in your web container (`docker-compose run web bash`, `./scripts/copydocs.py`) using [this work](https://openlibrary.org/works/OL16060982W/Prentice_Hall_Literature_-_Timeless_Voices_Timeless_Themes_-_The_American_Experience)
2. The error indicating the request is too large will be thrown (5495 chars. Limit says it's 4094)

* Actual: work is copied to local database
* Expected: error thrown due to data size

### Details

- **Logged in (Y/N)?**  N/A
- **Browser type/version?** N/A 
- **Operating system?** N/A
- **Environment (prod/dev/local)?** dev

### Related files

https://github.com/internetarchive/openlibrary/blob/master/scripts/copydocs.py

### PR DESCRIPTION
**Pull request by b2pacific**
Closes #5761 

fix

### Technical

Have set the size of http request line to 8190, had the option to set it to unlimited i.e 0. But don't think that this should be done.

### Stakeholders

@jimchamp

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -58,7 +58,7 @@ services:
     environment:
       - PYENV_VERSION=${PYENV_VERSION:-}
       - COVERSTORE_CONFIG=${COVERSTORE_CONFIG:-/openlibrary/conf/coverstore.yml}
-      - GUNICORN_OPTS=${GUNICORN_OPTS:- --reload --workers 1 --max-requests 250}
+      - GUNICORN_OPTS=${GUNICORN_OPTS:- --reload --workers 1 --max-requests 250 --limit-request-line 8190}
```
Comment (jimchamp) [inline docker-compose.yml:61]: The configuration looks correct, but this is the covers server.  Request that you revert this change, and add the `limit-request-line` flag to line 8 (the `services.web.environment` gunicorn configurations).

Comment (jimchamp) [reply docker-compose.yml:61]: You can test your changes with this URL (you'll have to restart Docker first):

http://localhost:8080/api/get_many?keys=[%22%2Fauthors%2FOL19725A%22%2C+%22%2Fauthors%2FOL47870A%22%2C+%22%2Fauthors%2FOL19512A%22%2C+%22%2Fauthors%2FOL265373A%22%2C+%22%2Fauthors%2FOL9252762A%22%2C+%22%2Fauthors%2FOL272057A%22%2C+%22%2Fauthors%2FOL115231A%22%2C+%22%2Fauthors%2FOL233050A%22%2C+%22%2Fauthors%2FOL130624A%22%2C+%22%2Fauthors%2FOL5854950A%22%2C+%22%2Fauthors%2FOL39316A%22%2C+%22%2Fauthors%2FOL646156A%22%2C+%22%2Fauthors%2FOL1816737A%22%2C+%22%2Fauthors%2FOL19158A%22%2C+%22%2Fauthors%2FOL118066A%22%2C+%22%2Fauthors%2FOL2400759A%22%2C+%22%2Fauthors%2FOL18405A%22%2C+%22%2Fauthors%2FOL3167773A%22%2C+%22%2Fauthors%2FOL116256A%22%2C+%22%2Fauthors%2FOL894149A%22%2C+%22%2Fauthors%2FOL29497A%22%2C+%22%2Fauthors%2FOL28035A%22%2C+%22%2Fauthors%2FOL391839A%22%2C+%22%2Fauthors%2FOL1272714A%22%2C+%22%2Fauthors%2FOL20188A%22%2C+%22%2Fauthors%2FOL242120A%22%2C+%22%2Fauthors%2FOL6971637A%22%2C+%22%2Fauthors%2FOL1129235A%22%2C+%22%2Fauthors%2FOL117830A%22%2C+%22%2Fauthors%2FOL719481A%22%2C+%22%2Fauthors%2FOL232501A%22%2C+%22%2Fauthors%2FOL29751A%22%2C+%22%2Fauthors%2FOL27342A%22%2C+%22%2Fauthors%2FOL28127A%22%2C+%22%2Fauthors%2FOL27349A%22%2C+%22%2Fauthors%2FOL2684942A%22%2C+%22%2Fauthors%2FOL2672053A%22%2C+%22%2Fauthors%2FOL763568A%22%2C+%22%2Fauthors%2FOL2434607A%22%2C+%22%2Fauthors%2FOL30968A%22%2C+%22%2Fauthors%2FOL113532A%22%2C+%22%2Fauthors%2FOL9252761A%22%2C+%22%2Fauthors%2FOL35183A%22%2C+%22%2Fauthors%2FOL241940A%22%2C+%22%2Fauthors%2FOL388126A%22%2C+%22%2Fauthors%2FOL22012A%22%2C+%22%2Fauthors%2FOL115504A%22%2C+%22%2Fauthors%2FOL351449A%22%2C+%22%2Fauthors%2FOL30250A%22%2C+%22%2Fauthors%2FOL2629213A%22%2C+%22%2Fauthors%2FOL19690A%22%2C+%22%2Fauthors%2FOL217730A%22%2C+%22%2Fauthors%2FOL2225458A%22%2C+%22%2Fauthors%2FOL43339A%22%2C+%22%2Fauthors%2FOL38897A%22%2C+%22%2Fauthors%2FOL2641167A%22%2C+%22%2Fauthors%2FOL528943A%22%2C+%22%2Fauthors%2FOL7023212A%22%2C+%22%2Fauthors%2FOL30224A%22%2C+%22%2Fauthors%2FOL27096A%22%2C+%22%2Fauthors%2FOL19804A%22%2C+%22%2Fauthors%2FOL870431A%22%2C+%22%2Fauthors%2FOL3290310A%22%2C+%22%2Fauthors%2FOL2629761A%22%2C+%22%2Fauthors%2FOL26681A%22%2C+%22%2Fauthors%2FOL2658871A%22%2C+%22%2Fauthors%2FOL26459A%22%2C+%22%2Fauthors%2FOL2746225A%22%2C+%22%2Fauthors%2FOL2877553A%22%2C+%22%2Fauthors%2FOL9252772A%22%2C+%22%2Fauthors%2FOL25582A%22%2C+%22%2Fauthors%2FOL27078A%22%2C+%22%2Fauthors%2FOL230834A%22%2C+%22%2Fauthors%2FOL151775A%22%2C+%22%2Fauthors%2FOL1390518A%22%2C+%22%2Fauthors%2FOL27318A%22%2C+%22%2Fauthors%2FOL32981A%22%2C+%22%2Fauthors%2FOL2677913A%22%2C+%22%2Fauthors%2FOL2623782A%22%2C+%22%2Fauthors%2FOL24527A%22%2C+%22%2Fauthors%2FOL22583A%22%2C+%22%2Fauthors%2FOL1234592A%22%2C+%22%2Fauthors%2FOL13640A%22%2C+%22%2Fauthors%2FOL20903A%22%2C+%22%2Fauthors%2FOL588695A%22%2C+%22%2Fauthors%2FOL30929A%22%2C+%22%2Fauthors%2FOL444165A%22%2C+%22%2Fauthors%2FOL19665A%22%2C+%22%2Fauthors%2FOL233499A%22%2C+%22%2Fauthors%2FOL399880A%22%2C+%22%2Fauthors%2FOL160545A%22%2C+%22%2Fauthors%2FOL709485A%22%2C+%22%2Fauthors%2FOL199956A%22%2C+%22%2Fauthors%2FOL18319A%22%2C+%22%2Fauthors%2FOL2652589A%22%2C+%22%2Fauthors%2FOL34361A%22%2C+%22%2Fauthors%2FOL27524A%22%2C+%22%2Fauthors%2FOL3620277A%22%2C+%22%2Fauthors%2FOL26069A%22%2C+%22%2Fauthors%2FOL25788A%22%2C+%22%2Fauthors%2FOL238257A%22%2C+%22%2Fauthors%2FOL23178A%22%2C+%22%2Fauthors%2FOL3245895A%22%2C+%22%2Fauthors%2FOL29394A%22%2C+%22%2Fauthors%2FOL1966702A%22%2C+%22%2Fauthors%2FOL766512A%22%2C+%22%2Fauthors%2FOL9252760A%22%2C+%22%2Fauthors%2FOL32584A%22%2C+%22%2Fauthors%2FOL4327652A%22%2C+%22%2Fauthors%2FOL60550A%22%2C+%22%2Fauthors%2FOL24762A%22%2C+%22%2Fauthors%2FOL836192A%22%2C+%22%2Fauthors%2FOL34170A%22%2C+%22%2Fauthors%2FOL21831A%22%2C+%22%2Fauthors%2FOL35145A%22%2C+%22%2Fauthors%2FOL31439A%22%2C+%22%2Fauthors%2FOL1232266A%22%2C+%22%2Fauthors%2FOL20925A%22%2C+%22%2Fauthors%2FOL25566A%22%2C+%22%2Fauthors%2FOL2746416A%22%2C+%22%2Fauthors%2FOL24457A%22%2C+%22%2Fauthors%2FOL1548599A%22%2C+%22%2Fauthors%2FOL3324313A%22%2C+%22%2Fauthors%2FOL9148252A%22%2C+%22%2Fauthors%2FOL30091A%22%2C+%22%2Fauthors%2FOL6841734A%22%2C+%22%2Fauthors%2FOL215971A%22%2C+%22%2Fauthors%2FOL220796A%22%2C+%22%2Fauthors%2FOL2630575A%22%2C+%22%2Fauthors%2FOL399682A%22%2C+%22%2Fauthors%2FOL219856A%22%2C+%22%2Fauthors%2FOL2769423A%22%2C+%22%2Fauthors%2FOL7572551A%22%2C+%22%2Fauthors%2FOL7720935A%22%2C+%22%2Fauthors%2FOL18040A%22%2C+%22%2Fauthors%2FOL2654524A%22%2C+%22%2Fauthors%2FOL424546A%22%2C+%22%2Fauthors%2FOL341986A%22%2C+%22%2Fauthors%2FOL20928A%22%2C+%22%2Fauthors%2FOL1476438A%22%2C+%22%2Fauthors%2FOL44633A%22%2C+%22%2Fauthors%2FOL33009A%22%2C+%22%2Fauthors%2FOL137220A%22%2C+%22%2Fauthors%2FOL2684610A%22%2C+%22%2Fauthors%2FOL392767A%22%2C+%22%2Fauthors%2FOL22420A%22%2C+%22%2Fauthors%2FOL7512434A%22%2C+%22%2Fauthors%2FOL26170A%22%2C+%22%2Fauthors%2FOL2657659A%22%2C+%22%2Fauthors%2FOL403331A%22%2C+%22%2Fauthors%2FOL21569A%22%2C+%22%2Fauthors%2FOL227752A%22%2C+%22%2Fauthors%2FOL30011A%22%2C+%22%2Fauthors%2FOL9256043A%22%2C+%22%2Fauthors%2FOL113959A%22%2C+%22%2Fauthors%2FOL233569A%22%2C+%22%2Fauthors%2FOL2690089A%22%2C+%22%2Fauthors%2FOL4435045A%22%2C+%22%2Fauthors%2FOL3292262A%22%2C+%22%2Fauthors%2FOL9252763A%22%2C+%22%2Fauthors%2FOL394640A%22%2C+%22%2Fauthors%2FOL221159A%22%2C+%22%2Fauthors%2FOL768231A%22%2C+%22%2Fauthors%2FOL532655A%22%2C+%22%2Fauthors%2FOL396367A%22%2C+%22%2Fauthors%2FOL114093A%22%2C+%22%2Fauthors%2FOL215841A%22%2C+%22%2Fauthors%2FOL545371A%22%2C+%22%2Fauthors%2FOL79124A%22%2C+%22%2Fauthors%2FOL32493A%22%2C+%22%2Fauthors%2FOL218453A%22]

Comment (jimchamp) [reply docker-compose.yml:61]: That will return a JSON string if your changes are correct. Otherwise, it will display an error message.

Comment (b2pacific) [reply docker-compose.yml:61]: Hey @jimchamp the limit-request-line argument dosen't seem to work. In the error message it's still showing that the limit is set to 4094, even after passing the argument!

Comment (jimchamp) [reply docker-compose.yml:61]: This works for me when the argument is passed for the web service, here: https://github.com/internetarchive/openlibrary/blob/d44e6e305e7149f48667fa829b6aa0260c5d20ff/docker-compose.yml#L3-L18

After making the change and restarting, I can see a `Recreating openlibrary_web_1` message in the console.  This is a good indicator that the change has occurred.

What OS are you using?  Maybe that's causing some issues?  I'm running Ubuntu.

**Comment by jimchamp** [comment]
@b2pacific I've made the title of this PR a bit more concise.  Have you joined our Slack?  If you have, feel free to contact me directly using `@Jim Champ` if you need any clarifications or anything.  Otherwise, I'll make sure that you get an invite.

**Comment by b2pacific** [comment]
@jimchamp currently I am not in OpenLibrary's Slack Community.

**Comment by b2pacific** [comment]
So, here are a few points that I wanted to convey.

I am testing whether the changes are working or not by running the below command

`docker-compose exec web bash`

And then

`./scripts/copyDocs.py <the url that you provided or the works Id of the work that was mentioned in the issue>`

Previously I was starting the services by running the command

`docker-compose up`

And this command did not recreated the docker containers when I made changes to the docker-compose file. So, I deleted them and created new ones manually, but still that did not produce the result that I wanted, I was still getting the error

`5459 > 4094`

which means that the limit for the size of request line hasn't changed.

So, then I saw that there was one docker-compose.override.yml file. So, I explicitly mentioned the name of docker-compose.yml file in the command to see if I got a different output.

`docker-compose -f docker-compose.yml up`

Using the above command did recreated the dockers when I made changes to the docker-compose.yml file. But still I am getting the same error as earlier.

Looks like I am stuck :(

PS - I am using Ubuntu

**Comment by jimchamp** [comment]
Thanks for the detailed response.  A couple of thoughts:

1. I'm not sure if the URL that I provided earlier will work with `copydocs`.  Regardless, you can simply follow the link to see if your changes have worked (it's fine that you are testing with `copydocs`, but it may be a bit slower than testing with your browser).
2. `docker-compose.yml` is the correct file to change.  No need to worry about the other `docker-compose` files.
3. As it is currently, this PR affects our covers server but not our web server.  Did you update line number 8 of `docker-compose.yml` before testing? That line contains the Gunicorn configs for our web server.
4. I only needed to run `docker-compose up` in order for the new configurations to take effect.

Of the above points, number 3 is the most important.  I know that I keep bringing it up, but did you make the config change on line 8?

**Comment by b2pacific** [comment]
Oh F!!!!!!

Testing this PR with the copydocs script, doesn't work on matter what changes I make. But going to the url that you provided does verifies the changes that it's working. It seems the copydocs script uses the covers service, maybe.

This means I have been testing this PR wrong till now : |

**Comment by cdrini** [comment]
Increasing the URL max length might be difficult, since there are a few nginx's at play on production. Instead, can we change wherever this GET request is made to a POST? And move whatever data is currently in the URL into the POST body? That should have no limit. Not sure what API exactly is being hit though, but wanted to propose that solution :)

**Comment by b2pacific** [comment]
@jimchamp any idea about this??
Or any suggestions?

**Comment by jimchamp** [comment]
`copydocs.py` is using `get_many`: https://github.com/internetarchive/openlibrary/blob/42b8040257a5ff0933fd6cd1d860ac0e760f446d/openlibrary/api.py#L115-L124

By changing line 117 to `if len(keys) > 100:`, I was able to make this work locally:
`./scripts/copydocs.py /works/OL16060982W --recursive`

@cdrini is 100 too small of a number for this function?   Maybe we can add an optional `max_length=500` argument to `get_many()`, and set this explicitly in copydocs?

**Comment by cdrini** [comment]
Oh, hmm, I wonder if we change the spot just below that to trigger a POST request instead of a GET; so

https://github.com/internetarchive/openlibrary/blob/42b8040257a5ff0933fd6cd1d860ac0e760f446d/openlibrary/api.py#L126-L128

So:

```py
    def _get_many(self, keys):
        response = self._request("/api/get_many", method='POST', data={"keys": json.dumps(keys)})
        return response.json()['result']
```

That might do the trick? Then we can dissolve the `_get_many` function entirely. I think downstream infogami can handle both GET/POST in this way, but would be best to test!

If that seems too difficult/doesn't work, 100 seems good :+1:

**Comment by jimchamp** [comment]
Adding `method="POST"` does not work.  I'm getting 403 errors with "Permission Denied" as a response when running `copydocs.py`.

@b2pacific can you add a `max_length` parameter to `get_many` in `/openlibrary/api.py`?  Afterwards, adding `max_length=100` to the below get_many call should fix everything:
https://github.com/internetarchive/openlibrary/blob/94961cb1d61a7e5e8832b01f0f4b83118a243f6a/scripts/copydocs.py#L234

After making these changes, running the following command should be successful:
`docker-compose exec -uroot web ./scripts/copydocs.py /works/OL16060982W --recursive`

**Comment by b2pacific** [comment]
@jimchamp Done :)

**Comment by b2pacific** [comment]
@jimchamp done :)
I had to modify the scripts/tests/test_copydocs.py file because a test was failing.

File: openlibrary/api.py [hunk 2]
```diff
@@ -112,10 +112,11 @@ def get(self, key, v=None):
         response = self._request(key + '.json', params={'v': v} if v else {})
         return unmarshal(response.json())
 
-    def get_many(self, keys):
-        """Get multiple documents in a single request as a dictionary."""
-        if len(keys) > 500:
-            # get in chunks of 500 to avoid crossing the URL length limit.
+    def get_many(self, keys, max_length = 100):
```
Comment (jimchamp) [inline openlibrary/api.py:115]:
```suggestion
    def get_many(self, keys, max_length=500):
```
The default max length should be the original value, `500`, for the sake of backwards compatibility.  Other consumers of `get_many` may require a max length of 500 to operate properly.

File: scripts/copydocs.py [hunk 3]
```diff
@@ -231,7 +231,7 @@ def copy(src, dest, keys, comment, recursive=False, saved=None, cache=None):
         cache = {}
 
     def get_many(keys):
-        docs = marshal(src.get_many(keys).values())
+        docs = marshal(src.get_many(keys, 100).values())
```
Comment (jimchamp) [inline scripts/copydocs.py:234]:
```suggestion
        docs = marshal(src.get_many(keys, max_length=100).values())
```
Let's include the keyword for clarity.

File: scripts/tests/test_copydocs.py [hunk 4]
```diff
@@ -39,7 +39,7 @@ def get(self, key, revision=None):
         else:
             return revisions.get(revision, None)
 
-    def get_many(self, keys):
+    def get_many(self, keys, max_length = 100):
```
Comment (jimchamp) [inline scripts/tests/test_copydocs.py:42]:
```suggestion
    def get_many(self, keys, max_length=500):
```
This `get_many` signature should match that of openlibrary/api.py's `get_many`.

**Review by jimchamp — CHANGES_REQUESTED** [review]
I'll be happy to merge this once the requested changes have been made.

**Comment by b2pacific** [comment]
@jimchamp  Done :)
Sorry, I had forgot to revert the initial commit of this PR.

File: openlibrary/api.py [hunk 5]
```diff
@@ -112,10 +112,11 @@ def get(self, key, v=None):
...
 
-    def get_many(self, keys):
-        """Get multiple documents in a single request as a dictionary."""
-        if len(keys) > 500:
-            # get in chunks of 500 to avoid crossing the URL length limit.
+    def get_many(self, keys, max_length = 500):
+        """Get multiple documents in a single request as a dictionary.
+        """
+        if len(keys) > max_length:
+            # get in chunks of 100 to avoid crossing the URL length limit.
```
Comment (jimchamp) [inline openlibrary/api.py:119]:
```suggestion
            # Process in batches to avoid crossing the URL length limit.
```

**Review by jimchamp — APPROVED** [review]
Lgtm!  Thanks @b2pacific!
